File: DataProcessing/TiingoApi.cpp
//
// Created by Mingcan Li on 4/7/23.
//
#include "DataSource.h"
#include <iostream>
#include <QtNetwork>
#include <unistd.h>

using namespace std;
using namespace hmdf;

class TiingoApi : public DataSource{
public:
    TiingoApi(string token) {
        this->token = token;
        string uri = url + "/api/test?token="+token;
        QNetworkRequest request((QUrl(QString::fromStdString(uri))));
        request.setRawHeader("Content-Type", "application/json");

        manager = new QNetworkAccessManager(this);
        connect(manager, SIGNAL(finished(QNetworkReply*)),this, SLOT(readReply(QNetworkReply*)));
        manager->get(request);
        int retries = 0;
        waitForData();
        if (dr) {
            if (rep.contains("success")) {
                qDebug() << "Success";
            } else {
                qDebug() << "Failed connection to server: " + QString::fromStdString(rep.toStdString());
                ::exit(0);
            }
        } else {
            qDebug() << "Connection Failed";
        }
        dr = false;
    }

    ~TiingoApi(){

    }

    bool setTickerSymbol(string symbol, vector<string> params) override {
        symbol = symbol;
        if(subDir == "" && params.size() < 1) {
            qDebug() << "params required";
            return false;
        } else {
            subDir = params[0];
        }
        this->symbol = symbol;

        //test connection
        //set uri and send request
        string uri = url + subDir + "prices?tickers="+symbol +"&token="+token;
        QNetworkRequest request((QUrl(QString::fromStdString(uri))));
        request.setRawHeader("Content-Type", "application/json");
        manager->get(request);

        //Handle data response
        waitForData();
        if (dr) {
            dr= false;
            if (rep.size() > 30) {
                qDebug() << "Token name set to " + QString::fromStdString(symbol);
                return true;
            } else {
                qDebug() << "Bad token name/dir";
                return false;
            }
        } else {
            qDebug() << "Connection failed";
            return false;
        }
    }

    void setSampleFreq(int min) override {
        sampleFreq = min;
    }

    void setHistoryLen(int days) override {
        historyLen = days;
    }

    bool dataReady() override {
        return dr;
    }

    TDF* getDataFrame() override {
        return nullptr;
    }

    int getCurrentPrice() override {
        return 0;
    }

    void readReply(QNetworkReply* reply) override {
        rep = reply->readAll();
        dr = true;
    }

    void startLive() override {
        thread();
    }


private:
    void thread (){
        time_t curr_t = std::time(NULL);
        //get history data before current time;
        QNetworkAccessManager* man = new QNetworkAccessManager(this);
        connect(man, SIGNAL(finished(QNetworkReply*)),this, SLOT(readReply(QNetworkReply*)));
        getHistoryData(curr_t, man);
    }

    /**
     * Be careful when debugging this function, wrong implementation could result in huge waste of api access
     * @param curr_t
     * @param man
     */
    void getHistoryData(time_t curr_t, QNetworkAccessManager* man) {
        struct tm gmtm = *gmtime(&curr_t);
        //get the current time
        int curr_sec = gmtm.tm_sec;
        int curr_min = gmtm.tm_min;
        int curr_hr = gmtm.tm_hour;
        //use the current time to offset the time to 0:0:0 UTC current date
        time_t offset_t = curr_t - curr_sec - curr_min*60 - curr_hr*3600;
        time_t startDate = dateConversion(offset_t, historyLen);
        //query maximum days with <=5000 items at a time until the current date at 0:0:0 UTC
        int maxQueryDays = sampleFreq * 5000 / (24*60);
        qDebug() << "maxDays: " << maxQueryDays;
        for (time_t queryStartDate = startDate ; queryStartDate < offset_t ; queryStartDate+=maxQueryDays*24*60*60) {
            time_t queryEndDate = queryStartDate + maxQueryDays*24*60*60 > offset_t ? offset_t : queryStartDate + maxQueryDays*24*60*60;
            struct tm gmtStart = *gmtime(&queryStartDate);
            struct tm gmtEnd = *gmtime(&queryEndDate);

            //Date format in YYYY-MM-DD
            string startT = to_string(1900+gmtStart.tm_year) + "-" + to_string(gmtStart.tm_mon + 1) + "-"+ to_string(gmtStart.tm_mday);
            string endT = to_string(1900+gmtEnd.tm_year) + "-" + to_string(gmtEnd.tm_mon+1) + "-"+ to_string(gmtEnd.tm_mday);

            //set uri and send request
            string uri = url + subDir + "prices?tickers="+symbol + "&startDate="+startT+"&endDate="+endT+"&resampleFreq="+
                    to_string(sampleFreq)+"min"+"&token="+token;
            qDebug() <<"uri: "<< QString::fromStdString(uri);
            QNetworkRequest request((QUrl(QString::fromStdString(uri))));
            request.setRawHeader("Content-Type", "application/json");
            man->get(request);
            qDebug() << "Query Date interval: "<< QString::fromStdString(startT) + "--" + QString::fromStdString(endT);
            QCoreApplication::processEvents();
            waitForData();
            if (dr) {
                parseData();
            } else {
                qDebug() << "Connection failed";
            }
        }

    }

    //the function subtracts days from current time
    time_t dateConversion(time_t curr, int days) {
        const time_t ONE_DAY = 24*60*60;
        time_t res = curr - ONE_DAY * days;
        return res;
    }

    void parseData() {
        //TODO codes to parse data into dataframe
        QJsonDocument json =  QJsonDocument::fromJson(rep);
        QJsonObject field = json.array().at(0).toObject();
        QJsonArray priceJsonArr = field["priceData"].toArray();
        qDebug()<<priceJsonArr;
        dr = false;
    }

    void waitForData() {
        int retries = 0;
        while (!dr && retries < 100) {
            QCoreApplication::processEvents();
            usleep(50000);
            retries++;
        }
    }

    atomic<bool> dr; //data ready, set on when fetched, set off when accessed
    atomic<int> historyLen, sampleFreq;
    string token, url  = "https://api.tiingo.com/", subDir = "", symbol;
    QNetworkAccessManager *manager;
    QByteArray rep;
};

File: DataSource/TiingoApi.py
import requests
from datetime import datetime, timedelta
import threading
from DataSource import DataSource
from time import sleep
import pandas as pd


class TiingoApi(DataSource):
    def __init__(self, token):
        self.token = token
        self.url = "https://api.tiingo.com"
        self.subDir = ""
        self.symbol = ""
        self.rep = ""
        self.dfr = False
        self.sampleFreq = 10
        self.historyLen = 30
        self.manager = requests.Session()
        self.manager.headers.update({'Content-Type': 'application/json'})
        self.df = pd.DataFrame()
        self.test_connection()

    def test_connection(self):
        uri = self.url + "/api/test?token=" + self.token
        try:
            response = self.manager.get(uri)
        except requests.exceptions.RequestException:
            print("connection failed")
            exit(0)
        if "success" in response.json()['message']:
            print("Success")
        else:
            print("Failed connection to server: " + str(response.json()))
            exit(0)

    def set_ticker_symbol(self, symbol, params):
        self.symbol = symbol
        if self.subDir == "" and len(params) < 1:
            print("params required")
            return False
        else:
            self.subDir = params[0]

        # test connection
        # set uri and send request
        uri = self.url + self.subDir + "prices?tickers=" + self.symbol + "&token=" + self.token
        response = self.manager.get(uri)
        if len(str(response.json())) >= 30:
            print("success")
            return True
        else:
            print("failed")
            return False

    def set_sample_freq(self, min):
        self.sampleFreq = min

    def set_history_len(self, days):
        self.historyLen = days

    def data_updated(self):
        return self.dfr

    def get_data_frame(self):
        return self.df

    def get_current_price(self):
        return 0

    def read_reply(self, reply):
        self.rep = reply.content.decode("utf-8")

    def start_live(self):
        t = threading.Thread(target=self.thread)
        t.start()

    def thread(self):
        curr_t = int(datetime.utcnow().timestamp())
        # get history data before current time
        man = requests.Session()
        man.headers.update({'Content-Type': 'application/json'})
        self.get_history_data(curr_t)

    # get history data until 0:0:0 UTC today
    def get_history_data(self, curr_t):
        gmtm = datetime.utcfromtimestamp(curr_t)
        # get the current time
        curr_sec = gmtm.second
        curr_min = gmtm.minute
        curr_hr = gmtm.hour
        # use the current time to offset the time to 0:0:0 UTC current date
        offset_t = curr_t - curr_sec - curr_min * 60 - curr_hr * 3600
        start_t = curr_t - self.historyLen * 86400
        # query maximum days with <=5000 items at a time until the current date at 0:0:0 UTC
        max_query_days = self.sampleFreq * 5000 // (24 * 60)
        print("maxDays: ", max_query_days, " max per request: ", (max_query_days * 24 * 60) // self.sampleFreq)
        first = True;
        for query_start_date in range(start_t, offset_t, max_query_days * 86400):
            query_end_date = query_start_date + max_query_days * 86400
            if query_end_date > offset_t:
                query_end_date = offset_t
            start_date_str = self.date_conversion(query_start_date, 0)
            end_date_str = self.date_conversion(query_end_date, 0)
            print("QueryDate:" + start_date_str + "--" + end_date_str)
            uri = self.url + self.subDir + "prices?startDate=" + start_date_str + "&endDate=" + end_date_str + \
                  "&resampleFreq=" + str(self.sampleFreq) + "min&tickers=" + self.symbol + "&token=" + self.token
            response = self.manager.get(uri)
            print(len(response.json()[0]["priceData"]))
            if first:
                self.parseData(response.json()[0]["priceData"])
                first = False
            else:
                self.parseData(response.json()[0]["priceData"][1:])
            self.df.to_csv("HistoricalData.csv")

    @staticmethod
    def date_conversion(t, delta):
        gmtm = datetime.utcfromtimestamp(t - delta * 86400)
        yr = str(gmtm.year)
        mn = str(gmtm.month).zfill(2)
        dy = str(gmtm.day).zfill(2)
        return yr + "-" + mn + "-" + dy

    def parseData(self, json):
        self.df = pd.concat([self.df, pd.DataFrame.from_dict(json)], ignore_index=True)
